### PR TITLE
C++: Add a new query for calling `c_str` on temporary objects

### PR DESCRIPTION
**Pull request by MathiasVP**
This PR adds a new query that looks for variations of:
```cpp
char* p = std::string("hello").c_str();
use(p);
```

MRVA results look really good. On the top 1000 projects I'm seeing 25 results, and they all look like TPs 🎉.

In the future I'd like to extend the query to support more lifetimes than just "temporary objects", but for now this is good enough, I think.

### CODE REVIEW

File: cpp/ql/test/query-tests/Security/CWE/CWE-416/semmle/tests/UseOfStringAfterLifetimeEnds/test.cpp [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+  std::vector<const char*> v1;
+  v1.push_back(std::string("hello").c_str()); // BAD
+
+  std::vector<S> v2;
+  v2.push_back({ std::string("hello").c_str() }); // BAD
+
+  S s5[] = { { std::string("hello").c_str() } }; // BAD
+
+	char c = std::string("hello").c_str()[0]; // GOOD
```
Comment (jketema) [inline cpp/ql/test/query-tests/Security/CWE/CWE-416/semmle/tests/UseOfStringAfterLifetimeEnds/test.cpp:185]: Indentation seems weird here.

File: cpp/ql/test/query-tests/Security/CWE/CWE-416/semmle/tests/UseOfStringAfterLifetimeEnds/test.cpp [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+struct S {
+  const char* s;
+};
+
+void call_by_value(S);
+void call_by_cref(const S&);
+
+void call(const char*);
+
+const char* test(bool b1, bool b2) {
```
Comment (jketema) [inline cpp/ql/test/query-tests/Security/CWE/CWE-416/semmle/tests/UseOfStringAfterLifetimeEnds/test.cpp:164]: Just for completeness can we add some "GOOD" test cases, where the string is constructed in a different expression then the one in which `c_str` is called?

**Review by jketema — COMMENTED** [review]
Two small comments for now.

File: cpp/ql/test/query-tests/Security/CWE/CWE-416/semmle/tests/UseOfStringAfterLifetimeEnds/test.cpp (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/test/query-tests/Security/CWE/CWE-416/semmle/tests/UseOfStringAfterLifetimeEnds/test.cpp]: Fixed in 62c432f3c76fed88ef1fd61f7be96f59bcf7bc6b.

File: cpp/ql/test/query-tests/Security/CWE/CWE-416/semmle/tests/UseOfStringAfterLifetimeEnds/test.cpp (hunk 2 above)
Comment (MathiasVP) [reply cpp/ql/test/query-tests/Security/CWE/CWE-416/semmle/tests/UseOfStringAfterLifetimeEnds/test.cpp]: Good idea. I've added some in 7b8d16469248706a0f9f7549d99ee9a0815b74de now

**Comment by geoffw0** [comment]
Real world results: I did a smaller MRVA run and tested on some local databases as well.  Found two results, both LGTM.  :+1:

File: cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEnds.ql [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+  )
+  or
+  isStoredInContainer(e)
+}
+
+from Call c
+where
+  outlivesFullExpr(c) and
+  not c.isFromUninstantiatedTemplate(_) and
+  c.getTarget() instanceof StdStringCStr and
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEnds.ql:95]: Yes, we can definitely do that 👍. I did a MRVA run earlier which didn't show any new results on the top 1000 projects, but that obviously doesn't mean that we shouldn't do it.

[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEnds.ql]: Fixed in 2b36ba33f0a29d932f0e376421e32b14145a1bdb. I had to add another case to `isTemporary` due to the small differences between `data` and `c_str`.

[the remark this answers is not in the document]
Comment (geoffw0) [reply cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEnds.ql]: Fantastic!

**Comment by geoffw0** [comment]
I'm happy with this PR and I think it's a great new query.  Looks like @jketema is happy too.  DCA LGTM.  I think we just need the docs team to take a look (I've taken the liberty of adding the `ready-for-doc-review` label).

**Comment by jketema** [comment]
I haven't looked at all the details, but I trust @geoffw0's judgement here.

**Comment by MathiasVP** [comment]
> (I've taken the liberty of adding the ready-for-doc-review label).

Thank you. I was indeed going to do that as soon as I had gotten a 👍 from the team.

File: cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEnds.qhelp [hunk 4]
```diff
@@ -0,0 +1,44 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+
+<overview>
+<p>Calling <code>c_str</code> on a <code>std::string</code> object returns a pointer to the underlying character array.
+However, if the <code>std::string</code> object is destroyed, then the pointer returned by <code>c_str</code> is no
+longer valid. If the pointer is used after the <code>std::string</code> object is destroyed, then the behavior is undefined.
```
Comment (felicitymay) [inline cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEnds.qhelp:9]: Suggest:

```suggestion
When the <code>std::string</code> object is destroyed, the pointer returned by <code>c_str</code> is no
longer valid. If the pointer is used after the <code>std::string</code> object is destroyed, then the behavior is undefined.
```

File: cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEnds.qhelp [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+<recommendation>
+<p>
+Ensure that the pointer returned by <code>c_str</code> does not outlive the underlying <code>std::string</code> object.
+</p>
+</recommendation>
+
+<example>
+<p>
+The following example concatenates two <code>std::string</code> objects, and then convert the resulting string to a
```
Comment (felicitymay) [inline cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEnds.qhelp:21]:
```suggestion
The following example concatenates two <code>std::string</code> objects, and then converts the resulting string to a
```

File: cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEndsBad.cpp [hunk 6]
```diff
@@ -0,0 +1,7 @@
+#include <string>
+void work(const char*);
+
```
Comment (felicitymay) [inline cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEndsBad.cpp:3]: We normally include a [comment here](https://codeql.github.com/codeql-query-help/cpp/cpp-dangerous-cin/). Something like this, but I'm sure you can write this more clearly:

```suggestion

// BAD: `c_str` is destroyed on return so `work` acts on a pointer to invalid memory

```

File: cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEndsGood.cpp [hunk 7]
```diff
@@ -0,0 +1,7 @@
+#include <string>
+void work(const char*);
+
```
Comment (felicitymay) [inline cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEndsGood.cpp:3]: Similarly, we should add a `GOOD` comment here.

**Review by felicitymay — DISMISSED** [review]
Thanks for the review request. This looks good. Just a couple of small comments.

File: cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEndsGood.cpp (hunk 7 above)
Comment (MathiasVP) [reply cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEndsGood.cpp:3]: Thanks! Added in 351caaccfe59ae1ba7d5bc73fdf9b90c38d3b6e5.

File: cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEndsBad.cpp (hunk 6 above)
Comment (felicitymay) [reply cpp/ql/src/Security/CWE/CWE-416/UseOfStringAfterLifetimeEndsBad.cpp:3]: Many thanks. When I checked the qhelp style guide, I was surprised to see this wasn't mentioned there. Indeed, the C qhelp files don't follow this convention consistently, but this is more consistent in other languages.

**Review by felicitymay — APPROVED** [review]
Thanks for the text updates. All LGTM.

**Pull request merged**